From f1614cca2400fcb82f9d782148419b82c25833cd Mon Sep 17 00:00:00 2001
From: Vishal Sagar <vishal.sagar@xilinx.com>
Date: Mon, 27 Jul 2020 00:49:32 -0700
Subject: [PATCH 1466/1852] drm: xlnx: mixer: handle software vblank

commit 1e438401adca9933b2aaf0869668ea90b7720209 from
https://github.com/Xilinx/linux-xlnx.git

During crtc begin/enable/disable, software vblank should
be on/off accordingly.

This fixes the kernel warning message below when running modetest for
same resolution second time but with different framerate.

setting mode 3840x2160-30Hz@BG24 on connectors 43, crtc 41
[   65.410018 ] ------------[ cut here  ]------------
[   65.414734 ] driver forgot to call drm_crtc_vblank_off()
[   65.420121 ] WARNING: CPU: 0 PID: 883 at
drivers/gpu/drm/drm_atomic_helper.c:1081
drm_atomic_helper_commit_modeset_disables+0x3f0/0x3f8
[   65.432457 ] Modules linked in: xilinx_hdmi_tx(O) xilinx_vphy(O)
dp159(O)
[   65.439305 ] CPU: 0 PID: 883 Comm: xmodetest Tainted: G           O
5.4.0-00001-g31f5c8934f7b-dirty #80
[   65.449254 ] Hardware name: ZynqMP ZCU102 Rev1.0 (DT)
[   65.454322 ] pstate: 60000005 (nZCv daif -PAN -UAO)
[   65.459214 ] pc :
drm_atomic_helper_commit_modeset_disables+0x3f0/0x3f8
[   65.465877 ] lr :
drm_atomic_helper_commit_modeset_disables+0x3f0/0x3f8
[   65.472541 ] sp : ffffffc0114639c0
[   65.475923 ] x29: ffffffc0114639c0 x28: ffffff8870afe800
[   65.481351 ] x27: ffffffc010fd9998 x26: 0000000000000000
[   65.486776 ] x25: 0000000000000038 x24: ffffffc011022741
[   65.492202 ] x23: ffffffc010dea648 x22: 0000000000000000
[   65.497628 ] x21: ffffff8870881800 x20: ffffff88714ad900
[   65.503053 ] x19: ffffff887931bd80 x18: 0000000000002d00
[   65.508479 ] x17: 0000000000000008 x16: 0000000000000000
[   65.513904 ] x15: 0000000000aaaaaa x14: 000000000000b400
[   65.519329 ] x13: 0000000000000020 x12: 00000000ffffffff
[   65.524755 ] x11: 0000000000000001 x10: ffffffc013e33000
[   65.530180 ] x9 : 00000000ffffffff x8 : 00000000000001e8
[   65.535606 ] x7 : ffffffc013e5d318 x6 : 00000000ffffffff
[   65.541031 ] x5 : ffffff8870bf83b0 x4 : 0000000000000000
[   65.546457 ] x3 : 0000000000000000 x2 : 00000000ffffffff
[   65.551883 ] x1 : ea77c02a8807f300 x0 : 0000000000000000
[   65.557308 ] Call trace:
[   65.559799 ]  drm_atomic_helper_commit_modeset_disables+0x3f0/0x3f8
[   65.566107 ]  drm_atomic_helper_commit_tail+0x20/0x78
[   65.571174 ]  commit_tail+0xe8/0x138
[   65.574735 ]  drm_atomic_helper_commit+0xc8/0x140
[   65.579450 ]  drm_atomic_commit+0x48/0x58
[   65.583452 ]  drm_atomic_helper_set_config+0xa0/0xb0
[   65.588432 ]  drm_mode_setcrtc+0x154/0x660
[   65.592526 ]  drm_ioctl_kernel+0xb8/0x108
[   65.596525 ]  drm_ioctl+0x200/0x410
[   65.599996 ]  do_vfs_ioctl+0x878/0xa20
[   65.603734 ]  ksys_ioctl+0x44/0x90
[   65.607111 ]  __arm64_sys_ioctl+0x1c/0x28
[   65.611121 ]  el0_svc_common.constprop.0+0x68/0x160
[   65.616008 ]  el0_svc_handler+0x6c/0x88
[   65.619830 ]  el0_svc+0x8/0xc
[   65.622768 ] ---[ end trace 88ff92a529061075  ]---

Signed-off-by: Vishal Sagar <vishal.sagar@xilinx.com>
Reviewed-by: Hyun Kwon <hyun.kwon@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/gpu/drm/xlnx/xlnx_mixer.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/drivers/gpu/drm/xlnx/xlnx_mixer.c b/drivers/gpu/drm/xlnx/xlnx_mixer.c
index dbb84b74e4f6..8621e53ac57b 100644
--- a/drivers/gpu/drm/xlnx/xlnx_mixer.c
+++ b/drivers/gpu/drm/xlnx/xlnx_mixer.c
@@ -2781,6 +2781,7 @@ xlnx_mix_crtc_atomic_disable(struct drm_crtc *crtc,
 {
 	xlnx_mix_crtc_dpms(crtc, DRM_MODE_DPMS_OFF);
 	xlnx_mix_clear_event(crtc);
+	drm_crtc_vblank_off(crtc);
 }
 
 static void xlnx_mix_crtc_mode_set_nofb(struct drm_crtc *crtc)
@@ -2797,6 +2798,7 @@ static void
 xlnx_mix_crtc_atomic_begin(struct drm_crtc *crtc,
 			   struct drm_crtc_state *old_crtc_state)
 {
+	drm_crtc_vblank_on(crtc);
 	/* Don't rely on vblank when disabling crtc */
 	if (crtc->state->event) {
 		struct xlnx_crtc *xcrtc = to_xlnx_crtc(crtc);
-- 
2.31.1

